R/spatial_nndm_cv.R

```diff
@@ -117,12 +118,41 @@ spatial_nndm_cv <- function(data, prediction_sites, ...,
   # we check both for length > 1 (in order to avoid the "condition has length"
   # error) and to see if the input is already only points
   pred_geometry <- unique(sf::st_geometry_type(prediction_sites))
-  if (length(pred_geometry) > 1 || pred_geometry != "POINT") {
-    prediction_sites <- sf::st_sample(
+
+  # these are more for clarity than control flow -- they do not get used
+  # outside of the below if/else
+  use_provided_points <- length(pred_geometry) == 1 && pred_geometry == "POINT"
+  sample_provided_poly <- length(pred_geometry) == 1 && pred_geometry %in% c(
+    "POLYGON",
+    "MULTIPOLYGON"
+  )
+  sample_bbox <- !use_provided_points && !sample_provided_poly
+
+  if (sample_bbox) {
+    sample_points <- sf::st_sample(
       x = sf::st_as_sfc(sf::st_bbox(prediction_sites)),
       size = prediction_sample_size,
       ...
     )
+  } else if (sample_provided_poly) {
+    sample_points <- sf::st_sample(
+      x = sf::st_geometry(prediction_sites),
+      size = prediction_sample_size,
+      ...
+    )
+  } else if (use_provided_points) {
+    sample_points <- prediction_sites
+  }
```

Comment:
I like it! 

Imo you can drop the comment about this being about clarity, I'd say that's a well-established pattern and doesn't need the explanation. 

I'd make the order of the ifelse statement the same as you laid it out above: sampling from the bounding box is the fall back so it shouldn't be the first condition to check. If the order is polygon > points > bbox, then that should be the order for both the definitions up in line 124-128 (no need for `sample_bbox` unless you want it) and the condition checking in the if-else block.

Reply:
Makes a ton of sense, thanks!